src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -199,11 +199,20 @@ public void FailRequest(Exception error)
 
             private void SetUrl()
             {
-                EventSourceTrace("Url: {0}", _requestMessage.RequestUri);
-                SetCurlOption(CURLoption.CURLOPT_URL, _requestMessage.RequestUri.AbsoluteUri);
+                string url = GetAbsoluteUri(_requestMessage.RequestUri);
+                EventSourceTrace("Url: {0}", url);
+                SetCurlOption(CURLoption.CURLOPT_URL, url);
                 SetCurlOption(CURLoption.CURLOPT_PROTOCOLS, (long)(CurlProtocols.CURLPROTO_HTTP | CurlProtocols.CURLPROTO_HTTPS));
             }
 
+            private static string GetAbsoluteUri(Uri uri)
+            {
+                // uri.AbsoluteUri/ToString() do not include IPv6 scope IDs, but when provided they
+                // should be used or else libcurl may be unable to correctly connect.  SerializationInfoString
+                // ensures these details are included.
+                return uri.GetComponents(UriComponents.SerializationInfoString, UriFormat.UriEscaped);
```

Comment:
/cc: @himadrisarkar 

+1 None of the System.Uri current methods except IdnHost will return a Punycode encoded string.
The best way to solve this is by implementing and testing a new `AbsoluteCanonicalUri` for Uri.

Alternatively, if the underlying stack will just pass-through the strings, you'll need to ensure that the host is Punycode encoded and the rest of the components are parsed according to [URI RFC](https://tools.ietf.org/html/rfc3986) semantics not [IRI RFC](https://tools.ietf.org/html/rfc3987). The current `System.Uri` default parser is going to return IRI semantics, which are not allowed within the HTTP protocol.